File: fastapi-python-backend/src/errors/app_error.py
from typing import Any

from ..types.api import ProblemDetails

ERROR_BASE_URI = 'https://api.example.dev/errors'


class AppError(Exception):
    def __init__(
        self,
        *,
        status: int,
        code: str,
        title: str,
        detail: str,
        type: str | None = None,
        is_operational: bool | None = None,
        errors: list[Any] | None = None,
        meta: dict[str, Any] | None = None,
        cause: Exception | None = None,
    ) -> None:
        super().__init__(detail)
        self.status = status
        self.code = code
        self.title = title
        self.detail = detail
        self.type = type or f'{ERROR_BASE_URI}/{code.lower().replace("_", "-")}'
        self.is_operational = is_operational if is_operational is not None else status < 500
        self.errors = errors or []
        self.meta = meta or {}

        if cause is not None:
            self.__cause__ = cause


def to_problem_details(
    error: AppError,
    *,
    instance: str,
    request_id: str | None = None,
    trace_id: str | None = None,
) -> ProblemDetails:
    return {
        'type': error.type,
        'title': error.title,
        'status': error.status,
        'detail': error.detail,
        'instance': instance,
        'code': error.code,
        'requestId': request_id,
        'traceId': trace_id,
        'errors': error.errors,
    }

File: hono-ts-bun-backend/src/errors/app-error.ts
import type { ProblemDetails } from "../types/api";

const ERROR_BASE_URI = "https://api.example.dev/errors";

export type AppErrorOptions = {
  status: number;
  code: string;
  title: string;
  detail: string;
  type?: string;
  isOperational?: boolean;
  errors?: unknown[];
  meta?: Record<string, unknown>;
  cause?: unknown;
};

export class AppError extends Error {
  status: number;
  code: string;
  title: string;
  detail: string;
  type: string;
  isOperational: boolean;
  errors: unknown[];
  meta: Record<string, unknown>;

  constructor(options: AppErrorOptions) {
    super(options.detail, options.cause ? { cause: options.cause } : undefined);
    this.name = "AppError";
    this.status = options.status;
    this.code = options.code;
    this.title = options.title;
    this.detail = options.detail;
    this.type =
      options.type ?? `${ERROR_BASE_URI}/${options.code.toLowerCase().replaceAll("_", "-")}`;
    this.isOperational = options.isOperational ?? options.status < 500;
    this.errors = options.errors ?? [];
    this.meta = options.meta ?? {};
  }
}

export function toProblemDetails(
  error: AppError,
  context: {
    instance: string;
    requestId?: string;
    traceId?: string;
  },
): ProblemDetails {
  return {
    type: error.type,
    title: error.title,
    status: error.status,
    detail: error.detail,
    instance: context.instance,
    code: error.code,
    requestId: context.requestId,
    traceId: context.traceId,
    errors: error.errors,
  };
}
